src/envs/screws.py

```diff
@@ -0,0 +1,482 @@
+        assert len(task.goal) == 1
+        goal_pred = list(task.goal)[0]
+        goal_screw = goal_pred.objects[0]
+
+        # Draw screws.
+        for screw in state.get_objects(self._screw_type):
+            screw_x = state.get(screw, "pose_x")
+            screw_y = state.get(screw, "pose_y")
+            if screw == goal_screw:
+                color = 'gold'
+            else:
+                color = 'grey'
+            rect = plt.Rectangle((screw_x - self._screw_width / 2.0,
+                                  screw_y - self._screw_height / 2.0),
+                                 self._screw_width,
+                                 self._screw_height,
+                                 edgecolor=color,
+                                 facecolor=color)
+            ax.add_patch(rect)
+
+        img = utils.fig2data(fig)
+        plt.close()
+        return [img]
+
+    @property
+    def predicates(self) -> Set[Predicate]:
+        return {
+            self._GripperCanPickScrew, self._AboveReceptacle,
+            self._HoldingScrew, self._ScrewInReceptacle
+        }
+
+    @property
+    def goal_predicates(self) -> Set[Predicate]:
+        return {self._ScrewInReceptacle}
+
+    @property
+    def types(self) -> Set[Type]:
+        return {self._screw_type, self._gripper_type, self._receptacle_type}
+
+    @property
+    def options(self) -> Set[ParameterizedOption]:
+        return {
+            self._MoveToScrew, self._MoveToReceptacle, self._MagnetizeGripper,
+            self._DemagnetizeGripper
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([-self._rz_width, -self._rz_height, 0.0],
+                          dtype=np.float32)
+        uppers = np.array([self._rz_width, self._rz_height, 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        dx, dy, magnetization = action.arr
+        if dx != 0.0 or dy != 0.0:
+            return self._transition_move(state, dx, dy)
+
+        if magnetization > 0.5:
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
+        screw_cluster_xcenters = []
+        for _ in range(self._num_screw_clusters):
+            screw_cluster_xcenters.append(
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
```

Comment:
should this be `rng.integers(1, possible_num_screws+1)`? (a) i don't think we'd want to allow for no screws and (b) we should be able to sample `possible_num_screws`, based on the name

Reply:
wait - `possible_num_screws` is a list that's passed in from `settings.py` (e.g. `[15, 20]`). I was intending this line to select a particular number from this list, which i believe is what it currently does

Reply:
oh sorry, my bad, ignore!